crypto/rsa/rsa_backend.c

```diff
@@ -64,22 +64,108 @@ static int collect_numbers(STACK_OF(BIGNUM) *numbers,
 int ossl_rsa_fromdata(RSA *rsa, const OSSL_PARAM params[], int include_private)
 {
     const OSSL_PARAM *param_n, *param_e,  *param_d = NULL;
-    BIGNUM *n = NULL, *e = NULL, *d = NULL;
+    const OSSL_PARAM *param_p, *param_q = NULL;
+    const OSSL_PARAM *param_derive = NULL;
+    BIGNUM *p = NULL, *q = NULL, *n = NULL, *e = NULL, *d = NULL;
     STACK_OF(BIGNUM) *factors = NULL, *exps = NULL, *coeffs = NULL;
     int is_private = 0;
+    int derive_from_pq = 0;
+    BN_CTX *ctx = NULL;
 
     if (rsa == NULL)
         return 0;
 
+    ctx = BN_CTX_new_ex(rsa->libctx);
+    if (ctx == NULL)
+        return 0;
+
+    param_derive = OSSL_PARAM_locate_const(params,
+                                           OSSL_PKEY_PARAM_RSA_DERIVE_FROM_PQ);
+    if ((param_derive != NULL)
+        && !OSSL_PARAM_get_int(param_derive, &derive_from_pq))
+        goto err;
+
     param_n = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_N);
     param_e = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_E);
     if (include_private)
         param_d = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_D);
 
-    if ((param_n != NULL && !OSSL_PARAM_get_BN(param_n, &n))
-        || (param_e != NULL && !OSSL_PARAM_get_BN(param_e, &e))
-        || (param_d != NULL && !OSSL_PARAM_get_BN(param_d, &d)))
-        goto err;
+    if (derive_from_pq) {
+        /* we need at minimum p, q and e */
+        param_p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_FACTOR1);
+        param_q = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_FACTOR2);
+        if ((param_p != NULL && !OSSL_PARAM_get_BN(param_p, &p))
+            || (param_q != NULL && !OSSL_PARAM_get_BN(param_q, &q))
+            || (param_e != NULL && !OSSL_PARAM_get_BN(param_e, &e))) {
+            ERR_raise(ERR_LIB_RSA, ERR_R_PASSED_NULL_PARAMETER);
+            goto err;
+        }
```

Comment:
This is wrong - this can be a multi=prime RSA and thus there might be additional factors. Of course in case of FIPS_MODULE it needs to be properly rejected in case they are there.

Reply:
If its a multiprime case, then this test still holds true, the first 2 factors are still required, as is e, no?

Rejection of the multiprime case when FIPS is enabled occurs lower down in the function, where we test the factors stack for having more than two primes.

I agree, this gets into a garbage-in, garbage out situation if someone passes in a set of paramters that include , say FACTOR1, FACTOR3 and FACTOR5 (similarly for the exponencts and coefficients list).  In the general case, things won't work if there are gaps in any of the parameter stack, but I think its reasonable to require that any parameters passed for which there are multiples be contiguous.

Thoughts?

Reply:
I mean n is a product of all the factors and not just p and q in case of multiprime rsa. Not that this particular condition is wrong - the whole computation of n and d is in the multiprime case. Perhaps just make it unsupported if there are more factors than 2? I do not think we care that much about this corner case.

Reply:
Thats true, perhaps in the multiprime case n and d should be required as p and q are.  I'll look at this after the multiprime derivation refactor

Reply:
corrected, we now still compute n in the 2 prime case, but check further down for the multiprime case, and, if n and d aren't provided issue an error